src/mds/MDSDmclockScheduler.cc

```diff
@@ -0,0 +1,891 @@
+    ceph_assert(default_conf.get_limit() == g_conf().get_val<double>("mds_dmclock_limit"));
+  }
+
+  /* need to check whether conf is updated from ceph.conf when the MDS is restarted */
+  dout(10) << __func__ <<  " enable " << default_conf.is_enabled()
+            << " reservation " << default_conf.get_reservation()
+            << " weight " << default_conf.get_weight()
+            << " limit " << default_conf.get_limit() << dendl;
+}
+
+mds_rank_t MDSDmclockScheduler::get_nodeid()
+{
+  if (mds != nullptr) {
+    return mds->get_nodeid();
+  }
+  return 0;
+}
+
+void MDSDmclockScheduler::set_mds_is_active(bool value)
+{
+  mds_is_active = value;
+}
+
+bool MDSDmclockScheduler::get_mds_is_active()
+{
+  return mds_is_active;
+}
+SessionMap *MDSDmclockScheduler::get_session_map()
+{
+  if (mds != nullptr) {
+    return &mds->sessionmap;
+  }
+  return 0;
+}
+
+void MDSDmclockScheduler::mds_lock()
+{
+  if (mds != nullptr) {
+    mds->mds_lock.lock();
+  }
+}
+
+void MDSDmclockScheduler::mds_unlock()
+{
+  if (mds != nullptr) {
+    mds->mds_lock.unlock();
+  }
+}
+
+int MDSDmclockScheduler::mds_is_locked_by_me()
+{
+  if (mds != nullptr) {
+    return ceph_mutex_is_locked_by_me(mds->mds_lock);
+  }
+  return 1;
+}
+
+std::string_view MDSDmclockScheduler::get_state_str() const
+{
+  switch(state) {
+    case SchedulerState::INIT:
+      return "INIT";
+    case SchedulerState::RUNNING:
+      return "RUNNING";
+    case SchedulerState::FINISHING:
+      return "FINISHING";
+    case SchedulerState::SHUTDOWN:
+      return "SHUTDOWN";
+    default:
+      return "UNKONWN";
+  }
+}
+
+Cost MDSDmclockScheduler::get_op_cost(const int op)
+{
+  switch (op) {
+    case CEPH_MDS_OP_LOOKUP:
+    case CEPH_MDS_OP_LOOKUPHASH:
+    case CEPH_MDS_OP_LOOKUPPARENT:
+    case CEPH_MDS_OP_LOOKUPINO:
+    case CEPH_MDS_OP_LOOKUPNAME:
+    case CEPH_MDS_OP_GETATTR:
+    case CEPH_MDS_OP_READDIR:
+    case CEPH_MDS_OP_OPEN:
+    case CEPH_MDS_OP_LOOKUPSNAP:
+    case CEPH_MDS_OP_LSSNAP:
+    case CEPH_MDS_OP_GETFILELOCK:
+      return 1;
+    case CEPH_MDS_OP_SETXATTR:
+    case CEPH_MDS_OP_SETATTR:
+    case CEPH_MDS_OP_RMXATTR:
+    case CEPH_MDS_OP_SETLAYOUT:
+    case CEPH_MDS_OP_SETDIRLAYOUT:
+    case CEPH_MDS_OP_MKNOD:
+    case CEPH_MDS_OP_LINK:
+    case CEPH_MDS_OP_UNLINK:
+    case CEPH_MDS_OP_RENAME:
+    case CEPH_MDS_OP_MKDIR:
+    case CEPH_MDS_OP_RMDIR:
+    case CEPH_MDS_OP_SYMLINK:
+    case CEPH_MDS_OP_CREATE:
+    case CEPH_MDS_OP_MKSNAP:
+    case CEPH_MDS_OP_RMSNAP:
+    case CEPH_MDS_OP_RENAMESNAP:
+    case CEPH_MDS_OP_SETFILELOCK:
+    default:
+      return 3;
+  }
+}
+
+bool MDSDmclockScheduler::_process_asok_qos_set(const cmdmap_t &cmdmap, std::ostream &ss)
+{
+  string path;
+  if(!cmd_getval(cmdmap, "path", path)) {
+    ss << "malformed path";
+    return false;
+  }
+
+  int64_t reservation;
+  if(!cmd_getval(cmdmap, "reservation", reservation) || reservation < 1) {
```

Comment:
i like the short code, but the lack clarity of the side-effect with reservation being set bugs me.
I'd prefer:
```
bool rc;
...

int64_t reservation;
rc = cmd_getval(cmdmap, "reservation", reservation);
if(!rc || reservation < 1) { ... }
```

Reply:
I will simplify the code.